File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        path = []

        def backtrack(i, target):
            if target == 0:
                result.append(path.copy())
                return
            
            if(i >= len(candidates)):return

            diff = target - candidates[i]

            if(diff >= 0):
                path.append(candidates[i])
                backtrack(i,diff)
                path.pop()

            backtrack(i+1, target)

        backtrack(0, target)
        return result

File: 39-combination-sum/combination-sum.java
class Solution {
    private List<List<Integer>> result = new ArrayList<List<Integer>>();
    private void backtrack(int[] candidates, int target, List<Integer> path, int start ){
        if(target == 0){
            // Collections.sort(path);
            // if(!result.contains(path))result.add(path);
            result.add(new ArrayList<>(path));
            return;
        }

        for (int i = start; i < candidates.length; i++){
            int diff = target - candidates[i];
            // if(diff >= 0){
            //     List<Integer> temp = new ArrayList<Integer>(path);
            //     temp.add(num);
            //     backtrack(candidates, diff, temp);
            // }
            if(diff >= 0){
                path.add(candidates[i]);
                backtrack(candidates, diff, path, i);
                path.remove(path.size()-1);
            }
        }
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        backtrack(candidates,target, new ArrayList<>(), 0);
        return result;
    }
}